### PR TITLE
Stored data sync

### PR DESCRIPTION
**Pull request by aljesusg**
@sergio-ocon Should we update the stored data of a charge only with OPEN pools?

### CODE REVIEW

File: app/models/manageiq/consumption/showback_charge.rb [hunk 1]
```diff
@@ -10,6 +10,7 @@ class ManageIQ::Consumption::ShowbackCharge < ApplicationRecord
   validates :showback_event, :presence => true, :allow_nil => false
 
   serialize :cost, JSON # Implement cost column as a JSON
```
Comment (chargio) [inline app/models/manageiq/consumption/showback_charge.rb:12]: Why do you serialize the cost? What would be the result of this. I believe this is a mistake on my side.

As you can see in the class, cost is cost_subunits and cost_currency, so it can be monetized. Adding a serialization wouldn't help (unless we want to store a list of costs, in which case I don't know how to store money in a JSON)

File: app/models/manageiq/consumption/showback_charge.rb [hunk 2]
```diff
@@ -18,7 +19,25 @@ def clean_cost
   end
 
   def stored_data_event
-    self.stored_data = self.showback_event.data
+    self.stored_data = {}
```
Comment (chargio) [inline app/models/manageiq/consumption/showback_charge.rb]: Shouldn't this be self.stored_data ||= {}
?
If you call this twice you will lose all data

File: app/models/manageiq/consumption/showback_charge.rb [hunk 3]
```diff
@@ -18,7 +19,25 @@ def clean_cost
...
+    self.stored_data[Time.now.utc] = self.showback_event.data
```
Comment (chargio) [inline app/models/manageiq/consumption/showback_charge.rb]: Wouldn't you need to be able to call this with the time you want?

In case the charge is done in the past to allow old charges to be done.

File: app/models/manageiq/consumption/showback_charge.rb [hunk 4]
```diff
@@ -18,7 +19,25 @@ def clean_cost
...
+  def stored_data_last
+    self.stored_data[stored_data_last_key]
+  end
+
+  def stored_data_last_key
+    self.stored_data.keys.sort.last
+  end
+
+  def update_stored_data
+    self.stored_data.delete(stored_data_last_key)
```
Comment (chargio) [inline app/models/manageiq/consumption/showback_charge.rb]: What happens if only stored_data_start is here?

**Review by chargio — APPROVED** [review]
Some comments.
I would need additional comments in code and docs :D

File: app/models/manageiq/consumption/showback_charge.rb (hunk 1 above)
Comment (aljesusg) [reply app/models/manageiq/consumption/showback_charge.rb:12]: I didn't change cost

**Comment by chargio** [comment]
UPDATED

Pool (and thus charges) should be in one of three states:

Open: 
You can change and update everything inside (add new charges, change numbers, etc)
Processing:
You can't add new charges, but you can update them charges (you are basically making the latest calculations). The charges should be up to date as a consequence of events no longer being updated or because they were updated and processed while creating the new pool
Closed:
No further changes


For charges, they should be in one of two states:
- Open
You can update them
- Closed
You can not update them

So basically the process should be:
You create one pool. Then you add and process changes.
At the end of the cycle, the pool changes into processing while the charges remain open. The final steps on processing are done, calculating the final rate. Charge will remain with starting data + end data.
Once all charges in the pool are closed, the pool changes to closed, no further changes are allowed.

I don't believe we need to maintain the state in the charge itself, as it is always accessed in the context of the pool, but this would be basically the process.

**Comment by chargio** [comment]
@aljesusg Answered your question

**Comment by chargio** [comment]
It should be possible to create charges so there is no need to go to the event to get rating

**Comment by aljesusg** [comment]
But I have some problems for make the migration because in your priceplan you call get measurements  for the event so...Should I look for pool of that event open and get any charge?

**Comment by aljesusg** [comment]
Something is worng here.. the specs is not getting the charge update...

Charge updated#<ManageIQ::Consumption::ShowbackCharge id: 62000000000564, cost_subunits: 0, cost_currency: "USD", showback_pool_id: 62000000000740, showback_event_id: 62000000001169, created_at: "2017-09-10 11:54:34", updated_at: "2017-09-10 11:54:34", stored_data: {"2017-09-10 06:54:34 UTC"=>{"CPU"=>{"average"=>[29.8571428571429, "percent"], "number"=>[2.0, "cores"], "max_number_of_cpu"=>[2, "cores"]}, "MEM"=>{"max_mem"=>[2048, "Mib"]}, "FLAVOR"=>{}}, 2017-09-10 11:54:34 UTC=>{"CPU"=>{"average"=>[49.8571428571429, "percent"], "number"=>[4.0, "cores"], "max_number_of_cpu"=>[6, "cores"]}, "MEM"=>{"max_mem"=>[2048, "Mib"]}, "FLAVOR"=>{}}}>
Charge result after update#<ManageIQ::Consumption::ShowbackCharge id: 62000000000564, cost_subunits: 0, cost_currency: "USD", showback_pool_id: 62000000000740, showback_event_id: 62000000001169, created_at: "2017-09-10 11:54:34", updated_at: "2017-09-10 11:54:34", stored_data: {"2017-09-10 06:54:34 UTC"=>{"CPU"=>{"average"=>[29.8571428571429, "percent"], "number"=>[2.0, "cores"], "max_number_of_cpu"=>[2, "cores"]}, "MEM"=>{"max_mem"=>[2048, "Mib"]}, "FLAVOR"=>{}}}>
